service/header/store_heightsub.go

```diff
@@ -0,0 +1,77 @@
+package header
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// heightSub provides a minimalistic mechanism to wait till header for a height becomes available.
+type heightSub struct {
+	height       uint64 // atomic
+	heightReqsLk sync.Mutex
+	heightReqs   map[uint64][]chan *ExtendedHeader
+}
+
+// newHeightSub instantiates new heightSub.
+func newHeightSub() *heightSub {
+	return &heightSub{
+		heightReqs: make(map[uint64][]chan *ExtendedHeader),
+	}
+}
+
+// Sub subscribes for a header of a given height.
+// It can return both values as nil, which means a requested header was already provided
+// and caller should get it elsewhere.
```

Comment:
Yes, from the store. In other words, it only subscribes for future heights and for known returns both nils.

Reply:
Agree that this could be more expressive. see also: https://github.com/celestiaorg/celestia-node/pull/428/files#r806313889
```go
if hs.Height() >= height {
	return nil, errElapsedHeight
}
```

Reply:
Changed to throw an error.